public/app/features/alerting/unified/hooks/useProduceNewRuleGroup.tsx

```diff
@@ -133,16 +97,196 @@ export function usePauseRuleInGroup() {
+      const targetNamespace = action.payload.namespaceName;
+
+      const oldGroupName = ruleGroup.groupName;
+      const targetGroupName = action.payload.groupName;
+
+      const isGroupRenamed = Boolean(targetGroupName) && oldGroupName !== targetGroupName;
+
+      // if we're also renaming the group, check if the target does not already exist
+      if (targetGroupName && isGroupRenamed) {
+        const targetGroup = await fetchRuleGroup({
+          rulerConfig,
+          namespace: targetNamespace,
+          group: targetGroupName,
+        })
+          .unwrap()
+          .catch(notFoundToNull);
+
+        if (targetGroup?.rules?.length) {
+          throw new Error('Target group already has rules, merging rule groups is currently not supported.');
+        }
+      }
+
+      // create the new group in the target namespace
+      await upsertRuleGroup({
+        rulerConfig,
+        namespace: targetNamespace,
+        payload: newRuleGroupDefinition,
+      }).unwrap();
+
+      const result = await deleteRuleGroup({
+        rulerConfig,
+        namespace: oldNamespace,
+        group: oldGroupName,
+      }).unwrap();
+
+      return result;
+    },
+    [produceNewRuleGroup, fetchRuleGroup, upsertRuleGroup, deleteRuleGroup]
+  );
+
+  return [moveFn, state] as const;
+}
+
+/**
+ * Rename a rule group but keep it within the same namespace, throws if the action targets an existing rule group
+ */
+export function useRenameRuleGroup() {
+  const [produceNewRuleGroup] = useProduceNewRuleGroup();
+  const [fetchRuleGroup] = alertRuleApi.endpoints.getRuleGroupForNamespace.useLazyQuery();
+  const [upsertRuleGroup] = alertRuleApi.endpoints.upsertRuleGroupForNamespace.useMutation();
+  const [deleteRuleGroup] = alertRuleApi.endpoints.deleteRuleGroupFromNamespace.useMutation();
+
+  const [state, renameFn] = useAsyncFn(
+    async (ruleGroup: RuleGroupIdentifier, groupName: string, interval?: string) => {
+      const action = renameRuleGroupAction({ groupName, interval });
+      const { newRuleGroupDefinition, rulerConfig } = await produceNewRuleGroup(ruleGroup, action);
+
+      const oldGroupName = ruleGroup.groupName;
+      const newGroupName = action.payload.groupName;
+      const namespaceName = ruleGroup.namespaceName;
+
+      // check if the target group exists
+      const targetGroup = await fetchRuleGroup({
+        rulerConfig,
+        namespace: namespaceName,
+        group: newGroupName,
+      })
+        .unwrap()
+        .catch(notFoundToNull);
+
+      if (targetGroup?.rules?.length) {
+        throw new Error('Target group has existing rules, merging rule groups is currently not supported.');
+      }
+
+      // if the target group does not exist, create the new group
+      const result = await upsertRuleGroup({
+        rulerConfig,
+        namespace: namespaceName,
+        payload: newRuleGroupDefinition,
+      }).unwrap();
+
+      // now delete the group we renamed
+      await deleteRuleGroup({
```

Comment:
This means that we now see "rules deleted" after the API completes, which might be scary (it worried me for a brief moment when I renamed a group to test)

Should we suppress some (or maybe all) of the messages within these hooks? And then we can provide one success message back to the user after all the requests have completed?

Reply:
Yea I think we may want to figure this out before merging – it currently does indeed reply with whatever the API responds with in the last requests we send.

I think perhaps we may want to suppress all toasts by default and manually show toasts? It'd increase the boilerplate a tad but I'm not too concerned about that. WDYT?

Reply:
I've appended some changes from https://github.com/grafana/grafana/pull/87435 and implemented that in the hooks so we can control when to render the success / error messages.